Podcast shows 0 episodes when add to Podcast App

The first podcast I attempted was Criminal. added ID to URL got an output.
On adding the output to my podcast app it creates an entry with the correct name and image but shows 0 episodes.

XML item example

```
<item>
<title>Protection | Criminal Podcast</title>
<description>John “Sonny” Franzese was once described as “largely responsible for the glamorization of the Mafia over the past century.” He’d been active in the Colombo crime family since the 1960s. And then, when he was 93, he was given an 8-year sentence. The evidence that helped convict him came from the last person he expected to wear a wire. Music in the episode “Tarte Tatin” by Blue Dot Sessions “Myrian” by Blue Dot Sessions “Story Four - Put it Down” by Blue Dot Sessions “Farsical - Slow Driver” by Blue Dot Sessions “Skeptic” by Podington Bear “Holo” by Blue Dot Sessions “In My Head” by Podington Bear “Sepia” by Podington Bear Criminal is a show about people who’ve done wrong, been wronged, or gotten caught somewhere in the middle. Hosted by Phoebe Judge. Part of the Vox Media Podcast Network. Follow the show on Apple Podcasts, Spotify, or wherever you get your podcasts. https://link.chtbl.com/n33u0HsA Say hello on Twitter (@CriminalShow), Instagram (@criminal_podcast), and Facebook https://www.facebook.com/ThisIsCriminal/. Sign up for our occasional newsletter, The Accomplice https://bit.ly/accomplicenewsletter. We also make This is Love and Phoebe Reads a Mystery. Artwork by Julienne Alexander http://www.yssrs.com/. Check out our online shop https://store.dftba.com/collections/criminal. Episode transcripts are posted on our website https://thisiscriminal.com/. Episode 105 originally published 12/21/2018 https://thisiscriminal.com/episode-105-protection-12-21-2018 #crime #podcast #truecrime</description>
<guid isPermaLink="false">dCN7FRPqj_4</guid>
<enclosure url="/media/youtube/dCN7FRPqj_4" length="0" type="audio/mp4"/>
<pubDate>Fri, 20 Jan 2023 03:34:09 +0000</pubDate>
<ns0:group xmlns:ns0="http://search.yahoo.com/mrss/">
<ns0:thumbnail url="https://i.ytimg.com/vi/dCN7FRPqj_4/maxresdefault.jpg"/>
</ns0:group>
</item>
```

On the PR when I enter a playlist ID on the URL I get
```
Internal Server Error

The server encountered an internal error and was unable to complete your request. Either the server is overloaded or there is an error in the application.
```

and on the logs I get
```

2024-02-01 22:37:27,157 [INFO] Generating RSS feed for YouTube playlist PLfikCHtGjo4kZ_kReUH9U3qCoaXQMwNDL
2024-02-01 22:37:27,297 [INFO] https://www.youtube.com/playlist?list=PLfikCHtGjo4kZ_kReUH9U3qCoaXQMwNDL
2024-02-01 22:37:27,298 [INFO] Grabbing episodes from YouTube playlist PLfikCHtGjo4kZ_kReUH9U3qCoaXQMwNDL
2024-02-01 22:37:27,495 [ERROR] Exception on /rss/youtube/PLfikCHtGjo4kZ_kReUH9U3qCoaXQMwNDL [GET]
Traceback (most recent call last):
  File "/home/appuser/.local/lib/python3.12/site-packages/flask/app.py", line 1463, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/appuser/.local/lib/python3.12/site-packages/flask/app.py", line 872, in full_dispatch_request
    rv = self.handle_user_exception(e)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/appuser/.local/lib/python3.12/site-packages/flask/app.py", line 870, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/appuser/.local/lib/python3.12/site-packages/flask/app.py", line 855, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/appuser/.local/lib/python3.12/site-packages/flask/views.py", line 110, in view
    return current_app.ensure_sync(self.dispatch_request)(**kwargs)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/appuser/.local/lib/python3.12/site-packages/flask/views.py", line 191, in dispatch_request
    return current_app.ensure_sync(meth)(**kwargs)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/src/podcastsponsorblock/views/youtuberssview.py", line 187, in get
    generate_rss_feed(episode_feed, GeneratorOptions(service_config, podcast_config, "")), mimetype="application/rss+xml"
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/appuser/.local/lib/python3.12/site-packages/cachetools/__init__.py", line 737, in wrapper
    v = func(*args, **kwargs)
        ^^^^^^^^^^^^^^^^^^^^^
  File "/app/src/podcastsponsorblock/views/youtuberssview.py", line 163, in generate_rss_feed
    feed_generator.add_entry(generate_episode_entry(episode, generator_options))
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/src/podcastsponsorblock/views/youtuberssview.py", line 93, in generate_episode_entry
    feed_entry.link(
  File "/home/appuser/.local/lib/python3.12/site-packages/feedgen/entry.py", line 445, in link
    self.__atom_link += ensure_format(
                        ^^^^^^^^^^^^^^
  File "/home/appuser/.local/lib/python3.12/site-packages/feedgen/util.py", line 74, in ensure_format
    raise ValueError('Data contains invalid keys')
ValueError: Data contains invalid keys
```

--- a/Dockerfile
+++ b/Dockerfile
@@ -21,6 +21,7 @@ CMD [ \
     "gunicorn", \
     "--log-level=info", \
     "--logger-class=podcastsponsorblock.AuthKeyFilteringLogger", \
+    "--workers=5", \
     "--log-file=-", \
     "--access-logfile=-", \
     "-b=0.0.0.0:8080", \

--- a/docs/configuration.md
+++ b/docs/configuration.md
@@ -1,5 +1,9 @@
 # Configuration
 
+Information on how to configure podcast-sponsor-block can be found below. Please read all sections as some podcast apps
+require special configuration.
+
+### Configuring the service
 podcast-sponsor-block is configured primarily via environment variables. A list of the available environment variables
 is below:
 
@@ -13,6 +17,41 @@ is below:
 | PODCAST_ALLOW_QUERY_PARAM_AUTH              | Allows `PODCAST_AUTH_KEY` to be provided as the query parameter `key`. Please note providing sensitive info as part of the query string [is bad practice](https://owasp.org/www-community/vulnerabilities/Information_exposure_through_query_strings_in_url), so only enable this option if your podcast app does not support HTTP basic authentication. When enabled, podcast-sponsor-block will do its best to redact the key from its logs. | No       | false         |
 | PODCAST_APPEND_AUTH_PARAM_TO_RESOURCE_LINKS | Causes the feed generator to include `key=<PODCAST_AUTH_KEY>` in resource links. This can only be enabled if `PODCAST_ALLOW_QUERY_PARAM_AUTH` is enabled, and is intended to be used only when it is required to workaround a lack of authentication support in your podcast app.                                                                                                                                                              | No       | false         |
 | PODCAST_TRUSTED_HOSTS                       | A comma-seperated list of trusted `Host` header values that can be used to access podcast-sponsor-block (e.g. `http://192.168.1.43:8081,https://podcasts.ericmedina024.com`). If configured, the `Host` header will be used to create absolute URLs. If not configured, relative URLs will be used instead which can cause issues with some podcast apps.                                                                                      | No       |               |
+
+### Configuring your podcasts
+
+Some podcast apps like Apple Podcasts require specifying additional attributes not available in the YouTube API. To use
+podcast-sponsor-block with these services, you will need to create a file called `podcasts.ini` within `PODCAST_DATA_PATH`.
+Within `podcasts.ini`, you can configure your podcasts like so (using the Sleep With Me podcast as an example):
+```ini
+# The name of the section is the YouTube playlist ID (not an alias)
+[PLMdYRoC0mZlW2uoesMXUrac26lsvOupSx]
+# the two-letter iso 639-1 language code. a list can be found here:
+# https://www.loc.gov/standards/iso639-2/php/code_list.php
+language=en
+
+# the iTunes ID of the podcast. this can be derived from the podcast url:
+# https://podcasts.apple.com/us/podcast/sleep-with-me/id740675898
+# the id is the numerical portion at the end of the url
+# Apple has very specific requirements for podcast thumbnails, so podcast-sponsor-block
+# will use this ID to grab a compliant thumbnail straight from iTunes
+itunes_id=740675898
+
+# the iTunes category this podcast fits in. there is a list of valid categories here:
+# https://podcasters.apple.com/support/1691-apple-podcasts-categories
+# please note: the category name IS case sensitive
+itunes_category=Leisure
+
+# If configured, this will override the YouTube playlist description
+description=Good night!
+
+# denotes whether this podcast contains explicit content (either yes or no)
+explicit=no
+```
+Please note that some podcast apps (like Apple Podcasts) require **all** of these options to be set. If you are having
+trouble with your podcast app not reading the RSS feed correctly, ensure you have set **all** these options for your
+podcast.
+
 
 ### Custom thumbnails
 Unfortunately, the YouTube API does not provide access to the cover image of podcasts. Due to this,

--- a/requirements.txt
+++ b/requirements.txt
@@ -5,3 +5,4 @@ python-dateutil
 feedgen==0.9.0 # v1.0 has a bug with link elements: https://github.com/lkiesow/python-feedgen/issues/135
 cachetools
 gunicorn
+requests

--- a/src/podcastsponsorblock/helpers/__init__.py
+++ b/src/podcastsponsorblock/helpers/__init__.py
@@ -1,4 +1,6 @@
 import re
+from xml.sax.saxutils import escape
+
 from .youtubeplaylistepisodefeed import YoutubePlaylistEpisodeFeed
 
 _LENIENT_YOUTUBE_ID_PATTERN = re.compile("^[A-Za-z0-9_-]{1,50}$")
@@ -8,4 +10,14 @@ def leniently_validate_youtube_id(potential_youtube_id: str) -> bool:
     return _LENIENT_YOUTUBE_ID_PATTERN.match(potential_youtube_id) is not None
 
 
-__all__ = ["YoutubePlaylistEpisodeFeed", "leniently_validate_youtube_id"]
+def escape_for_xml(unescaped_string: str):
+    return escape(unescaped_string, entities={
+        "'": "&apos;",
+        '"': "&quot;",
+        "©": "&#xA9;",
+        "℗": "&#x2117;",
+        "™": "&#x2122;"
+    })
+
+
+__all__ = ["YoutubePlaylistEpisodeFeed", "leniently_validate_youtube_id", "escape_for_xml"]

--- a/src/podcastsponsorblock/helpers/__init__.py
+++ b/src/podcastsponsorblock/helpers/__init__.py
@@ -1,9 +1,39 @@
 import re
+from datetime import timedelta
+from urllib.parse import urlparse
 from xml.sax.saxutils import escape
+
+import requests
+from cachetools import cached, TTLCache
 
 from .youtubeplaylistepisodefeed import YoutubePlaylistEpisodeFeed
 
 _LENIENT_YOUTUBE_ID_PATTERN = re.compile("^[A-Za-z0-9_-]{1,50}$")
+
+
+def transform_artwork_url(artwork_url: str, new_height: int, new_width: int) -> str:
+    parsed_url = urlparse(artwork_url)
+    split_path = parsed_url.path.split("/")
+    split_path[len(split_path) - 1] = f"{new_height}x{new_width}bb.jpg"
+    new_size_path = "/".join(split_path)
+    return parsed_url._replace(path=new_size_path).geturl()
+
+
+@cached(cache=TTLCache(ttl=timedelta(minutes=60).total_seconds(), maxsize=1024))
+def get_itunes_artwork(itunes_id: str) -> str:
+    itunes_response = requests.get("https://itunes.apple.com/lookup?id=", params={"id": itunes_id})
+    if itunes_response.status_code != 200:
+        raise ValueError(f"Invalid iTunes ID: {itunes_id}")
+    itunes_results = itunes_response.json().get("results", tuple())
+    if len(itunes_results) != 1:
+        raise ValueError(f"{len(itunes_results)} results found for iTunes ID: {itunes_id}")
+    matching_podcast = itunes_results[0]
+    artwork_key = next((key for key in matching_podcast.keys() if key.startswith("artworkUrl")), None)
+    if artwork_key is None:
+        raise ValueError(f"iTunes ID missing artwork: {itunes_id}")
+    # even if a 3000x3000 artwork isn't available, it will return the highest available resolution. 3000x3000 is the max
+    # allowed
+    return transform_artwork_url(matching_podcast[artwork_key], new_height=3000, new_width=3000)
 
 
 def leniently_validate_youtube_id(potential_youtube_id: str) -> bool:
@@ -20,4 +50,4 @@ def escape_for_xml(unescaped_string: str):
     })
 
 
-__all__ = ["YoutubePlaylistEpisodeFeed", "leniently_validate_youtube_id", "escape_for_xml"]
+__all__ = ["YoutubePlaylistEpisodeFeed", "leniently_validate_youtube_id", "escape_for_xml", "get_itunes_artwork"]

--- a/src/podcastsponsorblock/helpers/youtubeplaylistepisodefeed.py
+++ b/src/podcastsponsorblock/helpers/youtubeplaylistepisodefeed.py
@@ -9,7 +9,7 @@ from dateutil.parser import isoparse as parse_iso_date
 from flask import url_for
 
 from .. import views
-from ..models import ItemDetails, EpisodeDetails, Author, ServiceConfig
+from ..models import ItemDetails, EpisodeDetails, Author, FeedOptions
 
 from googleapiclient.discovery import build as build_google_api_client
 
@@ -132,32 +132,32 @@ def get_episodes_cached(
 )
 def get_logo_cached(
     youtube_client: "YoutubeClient",
-    config: ServiceConfig,
+    feed_options: FeedOptions,
     playlist_details: ItemDetails,
 ) -> str:
-    thumbnail_path = views.get_thumbnail_path(playlist_details.id, config)
+    thumbnail_path = views.get_thumbnail_path(playlist_details.id, feed_options)
     if thumbnail_path is None:
         channel_details = get_channel_details(
             youtube_client, playlist_details.author.id
         )
         return channel_details.icon_url
     else:
-        if config.append_auth_param_to_resource_links:
+        if feed_options.service_config.append_auth_param_to_resource_links:
             return url_for(
                 "thumbnail_view",
                 thumbnail_key=playlist_details.id,
-                key=config.auth_key,
+                key=feed_options.service_config.auth_key,
             )
         return url_for("thumbnail_view", thumbnail_key=playlist_details.id)
 
 
 class YoutubePlaylistEpisodeFeed:
-    def __init__(self, playlist_id: str, config: ServiceConfig):
-        self.config = config
+    def __init__(self, playlist_id: str, feed_options: FeedOptions):
+        self.feed_options = feed_options
         self.youtube_client = build_google_api_client(
             "youtube",
             "v3",
-            developerKey=self.config.youtube_api_key,
+            developerKey=self.feed_options.service_config.youtube_api_key,
             cache_discovery=False,
         )
         self.playlist_details = get_playlist_details(self.youtube_client, playlist_id)
@@ -166,7 +166,7 @@ class YoutubePlaylistEpisodeFeed:
 
     @property
     def logo(self) -> str:
-        return get_logo_cached(self.youtube_client, self.config, self.playlist_details)
+        return get_logo_cached(self.youtube_client, self.feed_options, self.playlist_details)
 
     @property
     def episodes(self) -> Sequence[EpisodeDetails]:

--- a/src/podcastsponsorblock/helpers/youtubeplaylistepisodefeed.py
+++ b/src/podcastsponsorblock/helpers/youtubeplaylistepisodefeed.py
@@ -1,7 +1,7 @@
 import logging
 from datetime import timedelta
 from operator import attrgetter
-from typing import Iterable, Optional, Sequence, TYPE_CHECKING
+from typing import Iterable, Optional, Sequence, TYPE_CHECKING, Any, Callable, Hashable
 
 from cachetools import cached, TTLCache
 from cachetools.keys import hashkey
@@ -90,6 +90,11 @@ def remove_unavailable_items(playlist_items: Sequence[dict]) -> Sequence[dict]:
     )
 
 
+def remove_duplicates(objects: Sequence[Any], key_getter: Callable[[Any], Hashable]) -> Sequence[Any]:
+    unique_object_dict = {key_getter(obj): obj for obj in objects}
+    return tuple(unique_object_dict.values())
+
+
 @cached(
     TTLCache(maxsize=1024, ttl=timedelta(minutes=60).total_seconds()),
     key=lambda _, playlist_details: hashkey(playlist_details.id),
@@ -112,10 +117,11 @@ def get_episodes_cached(
             playlist_items_request, playlist_items_response
         )
         continue_requesting_playlist_items = playlist_items_request is not None
-    return sorted(
+    sorted_playlist_episodes = sorted(
         map(create_episode_details, remove_unavailable_items(all_playlist_items)),
         key=attrgetter("published_at"),
     )
+    return remove_duplicates(sorted_playlist_episodes, attrgetter("id"))
 
 
 @cached(

--- a/src/podcastsponsorblock/helpers/youtubeplaylistepisodefeed.py
+++ b/src/podcastsponsorblock/helpers/youtubeplaylistepisodefeed.py
@@ -9,7 +9,7 @@ from dateutil.parser import isoparse as parse_iso_date
 from flask import url_for
 
 from .. import views
-from ..models import ItemDetails, EpisodeDetails, Author, Configuration
+from ..models import ItemDetails, EpisodeDetails, Author, ServiceConfig
 
 from googleapiclient.discovery import build as build_google_api_client
 
@@ -90,7 +90,9 @@ def remove_unavailable_items(playlist_items: Sequence[dict]) -> Sequence[dict]:
     )
 
 
-def remove_duplicates(objects: Sequence[Any], key_getter: Callable[[Any], Hashable]) -> Sequence[Any]:
+def remove_duplicates(
+    objects: Sequence[Any], key_getter: Callable[[Any], Hashable]
+) -> Sequence[Any]:
     unique_object_dict = {key_getter(obj): obj for obj in objects}
     return tuple(unique_object_dict.values())
 
@@ -130,7 +132,7 @@ def get_episodes_cached(
 )
 def get_logo_cached(
     youtube_client: "YoutubeClient",
-    config: Configuration,
+    config: ServiceConfig,
     playlist_details: ItemDetails,
 ) -> str:
     thumbnail_path = views.get_thumbnail_path(playlist_details.id, config)
@@ -150,7 +152,7 @@ def get_logo_cached(
 
 
 class YoutubePlaylistEpisodeFeed:
-    def __init__(self, playlist_id: str, config: Configuration):
+    def __init__(self, playlist_id: str, config: ServiceConfig):
         self.config = config
         self.youtube_client = build_google_api_client(
             "youtube",

--- a/src/podcastsponsorblock/main.py
+++ b/src/podcastsponsorblock/main.py
@@ -16,7 +16,7 @@ def initialize_authorization(
     def require_authentication():
         if (
             allow_query_param_auth
-            and request.method == "GET"
+            and request.args is not None
             and request.args.get("key") == key
         ):
             return

--- a/src/podcastsponsorblock/main.py
+++ b/src/podcastsponsorblock/main.py
@@ -1,11 +1,13 @@
+import json
 import logging
 import os
+from configparser import ConfigParser
 from pathlib import Path
 from typing import Optional, MutableMapping, Sequence
 
 from flask import Flask, request, Response, Request
 
-from .models import Configuration
+from .models import ServiceConfig, PodcastConfig
 from .views import YoutubeMediaView, YoutubeRSSView, ThumbnailView
 
 
@@ -51,12 +53,31 @@ def parse_comma_seperated_value(hostname_str: Optional[str]) -> Sequence[str]:
     return tuple()
 
 
-def populate_config(source: MutableMapping) -> Configuration:
+def parse_podcast_configs(config_path: Path) -> dict[str, PodcastConfig]:
+    if not config_path.exists() or not config_path.is_file():
+        return dict()
+    config = ConfigParser()
+    config.read(config_path)
+    podcast_configs = dict()
+    for section_name in config.sections():
+        section_values = config[section_name]
+        podcast_configs[section_name] = PodcastConfig(
+            id=section_name,
+            language=section_values.get("language"),
+            description=section_values.get("description"),
+            itunes_category=section_values.get("itunes_category"),
+            explicit=section_values.getboolean("explicit"),
+        )
+    return podcast_configs
+
+
+def populate_service_config(source: MutableMapping) -> ServiceConfig:
+    data_path = Path(source["PODCAST_DATA_PATH"]).absolute().resolve()
     try:
-        return Configuration(
+        return ServiceConfig(
             youtube_api_key=source.pop("PODCAST_YOUTUBE_API_KEY"),
             auth_key=source.pop("PODCAST_AUTH_KEY", None),
-            data_path=Path(source["PODCAST_DATA_PATH"]).absolute().resolve(),
+            data_path=data_path,
             allow_query_param_auth=is_true(
                 source.get("PODCAST_ALLOW_QUERY_PARAM_AUTH", None)
             ),
@@ -70,18 +91,20 @@ def populate_config(source: MutableMapping) -> Configuration:
             trusted_hosts=parse_comma_seperated_value(
                 source.get("PODCAST_TRUSTED_HOSTS", None)
             ),
+            podcast_configs=parse_podcast_configs(data_path / "podcasts.ini")
         )
     except KeyError as exception:
         # noinspection PyUnresolvedReferences
         raise ValueError(f"Missing configuration value: {exception}")
 
 
-def log_config(config: Configuration) -> None:
+def log_service_config(config: ServiceConfig) -> None:
     logging.info(f"Loaded configuration:")
     logging.info(f"  - Data path: {config.data_path}")
     logging.info(f"  - Trusted hosts: {config.trusted_hosts}")
     logging.info(f"  - Aliases: {config.aliases}")
     logging.info(f"  - Categories to remove: {config.categories_to_remove}")
+    logging.info(f"   - Podcast configs: {config.podcast_configs}")
     logging.info(
         f"  - YouTube key: {'(configured)' if config.auth_key is not None else ''}"
     )
@@ -100,13 +123,13 @@ def create_app() -> Flask:
     )
     logging.info("Creating app")
     app = Flask(__name__)
-    config = populate_config(os.environ)
+    config = populate_service_config(os.environ)
     if not config.allow_query_param_auth and config.append_auth_param_to_resource_links:
         raise ValueError(
             "Cannot append auth param to resource links when query auth is not allowed"
         )
-    log_config(config)
-    app.config["PODCAST_CONFIG"] = config
+    log_service_config(config)
+    app.config["PODCAST_SERVICE_CONFIG"] = config
     if config.allow_query_param_auth:
         from . import AuthKeyFilteringLogger
 

--- a/src/podcastsponsorblock/main.py
+++ b/src/podcastsponsorblock/main.py
@@ -67,6 +67,7 @@ def parse_podcast_configs(config_path: Path) -> dict[str, PodcastConfig]:
             description=section_values.get("description"),
             itunes_category=section_values.get("itunes_category"),
             explicit=section_values.getboolean("explicit"),
+            itunes_id=section_values.get("itunes_id")
         )
     return podcast_configs
 

--- a/src/podcastsponsorblock/models/__init__.py
+++ b/src/podcastsponsorblock/models/__init__.py
@@ -11,6 +11,7 @@ class PodcastConfig:
     description: Optional[str]
     itunes_category: Optional[str]
     explicit: Optional[bool]
+    itunes_id: Optional[str]
 
 
 @dataclass
@@ -24,6 +25,13 @@ class ServiceConfig:
     categories_to_remove: Sequence[str]
     trusted_hosts: Sequence[str]
     podcast_configs: dict[str, PodcastConfig]
+
+
+@dataclass
+class FeedOptions:
+    service_config: ServiceConfig
+    podcast_config: Optional[PodcastConfig]
+    host: str
 
 
 @dataclass

--- a/src/podcastsponsorblock/models/__init__.py
+++ b/src/podcastsponsorblock/models/__init__.py
@@ -5,7 +5,16 @@ from typing import Optional, Sequence
 
 
 @dataclass
-class Configuration:
+class PodcastConfig:
+    id: str
+    language: Optional[str]
+    description: Optional[str]
+    itunes_category: Optional[str]
+    explicit: Optional[bool]
+
+
+@dataclass
+class ServiceConfig:
     youtube_api_key: str
     auth_key: Optional[str]
     data_path: Path
@@ -14,6 +23,7 @@ class Configuration:
     aliases: dict[str, str]
     categories_to_remove: Sequence[str]
     trusted_hosts: Sequence[str]
+    podcast_configs: dict[str, PodcastConfig]
 
 
 @dataclass
@@ -26,7 +36,7 @@ class Author:
 class ItemDetails:
     id: str
     title: str
-    description: str
+    description: Optional[str]
     author: Author
     icon_url: str
 

--- a/src/podcastsponsorblock/views/thumbnailview.py
+++ b/src/podcastsponsorblock/views/thumbnailview.py
@@ -5,7 +5,7 @@ from flask import current_app, send_file, Response
 from flask.typing import ResponseReturnValue
 from flask.views import MethodView
 
-from ..models import ServiceConfig
+from ..models import FeedOptions
 
 
 def compute_potential_thumbnail_stems(
@@ -20,12 +20,12 @@ def compute_potential_thumbnail_stems(
     return tuple(name.casefold() for name in all_potential_thumbnail_names)
 
 
-def get_thumbnail_path(thumbnail_key: str, config: ServiceConfig) -> Optional[Path]:
-    thumbnail_directory = config.data_path / "thumbnails"
+def get_thumbnail_path(thumbnail_key: str, feed_options: FeedOptions) -> Optional[Path]:
+    thumbnail_directory = feed_options.service_config.data_path / "thumbnails"
     if not thumbnail_directory.exists() or not thumbnail_directory.is_dir():
         return None
     all_potential_thumbnail_stems = compute_potential_thumbnail_stems(
-        thumbnail_key, config.aliases
+        thumbnail_key, feed_options.service_config.aliases
     )
     for candidate_thumbnail in thumbnail_directory.iterdir():
         if (

--- a/src/podcastsponsorblock/views/thumbnailview.py
+++ b/src/podcastsponsorblock/views/thumbnailview.py
@@ -1,4 +1,3 @@
-import logging
 from pathlib import Path
 from typing import Optional, Sequence
 
@@ -46,3 +45,5 @@ class ThumbnailView(MethodView):
         if thumbnail_path is None:
             return Response("Thumbnail not found", status=404)
         return send_file(thumbnail_path)
+
+    head = get

--- a/src/podcastsponsorblock/views/thumbnailview.py
+++ b/src/podcastsponsorblock/views/thumbnailview.py
@@ -46,4 +46,5 @@ class ThumbnailView(MethodView):
             return Response("Thumbnail not found", status=404)
         return send_file(thumbnail_path)
 
-    head = get
+    def head(self, thumbnail_key: str) -> ResponseReturnValue:
+        return self.get(thumbnail_key)

--- a/src/podcastsponsorblock/views/thumbnailview.py
+++ b/src/podcastsponsorblock/views/thumbnailview.py
@@ -5,7 +5,7 @@ from flask import current_app, send_file, Response
 from flask.typing import ResponseReturnValue
 from flask.views import MethodView
 
-from ..models import Configuration
+from ..models import ServiceConfig
 
 
 def compute_potential_thumbnail_stems(
@@ -20,9 +20,9 @@ def compute_potential_thumbnail_stems(
     return tuple(name.casefold() for name in all_potential_thumbnail_names)
 
 
-def get_thumbnail_path(thumbnail_key: str, config: Configuration) -> Optional[Path]:
+def get_thumbnail_path(thumbnail_key: str, config: ServiceConfig) -> Optional[Path]:
     thumbnail_directory = config.data_path / "thumbnails"
-    if not thumbnail_directory.exists():
+    if not thumbnail_directory.exists() or not thumbnail_directory.is_dir():
         return None
     all_potential_thumbnail_stems = compute_potential_thumbnail_stems(
         thumbnail_key, config.aliases
@@ -40,7 +40,7 @@ def get_thumbnail_path(thumbnail_key: str, config: Configuration) -> Optional[Pa
 class ThumbnailView(MethodView):
     def get(self, thumbnail_key: str) -> ResponseReturnValue:
         thumbnail_path = get_thumbnail_path(
-            thumbnail_key, current_app.config["PODCAST_CONFIG"]
+            thumbnail_key, current_app.config["PODCAST_SERVICE_CONFIG"]
         )
         if thumbnail_path is None:
             return Response("Thumbnail not found", status=404)

--- a/src/podcastsponsorblock/views/youtubemediaview.py
+++ b/src/podcastsponsorblock/views/youtubemediaview.py
@@ -52,6 +52,8 @@ class YoutubeMediaView(MethodView):
         self.video_download_locks = defaultdict(threading.Lock)
 
     def get(self, video_id: str) -> ResponseReturnValue:
+        # Apple podcasts requires the file extension, but we don't want it and need to remove it if present
+        video_id = video_id.removesuffix(".m4a")
         if not leniently_validate_youtube_id(video_id):
             return Response("Invalid video ID", status=400)
         config: ServiceConfig = current_app.config["PODCAST_SERVICE_CONFIG"]

--- a/src/podcastsponsorblock/views/youtubemediaview.py
+++ b/src/podcastsponsorblock/views/youtubemediaview.py
@@ -13,7 +13,7 @@ from ..helpers import leniently_validate_youtube_id
 from googleapiclient.discovery import build as build_google_api_client
 import threading
 
-from ..models import Configuration
+from ..models import ServiceConfig
 
 
 def download_m4a_audio(
@@ -35,7 +35,7 @@ def download_m4a_audio(
         youtube_dlp_client.download((f"https://www.youtube.com/watch?v={video_id}",))
 
 
-def validate_youtube_video_id(video_id: str, config: Configuration) -> Optional[str]:
+def validate_youtube_video_id(video_id: str, config: ServiceConfig) -> Optional[str]:
     youtube_client = build_google_api_client(
         "youtube", "v3", developerKey=config.youtube_api_key, cache_discovery=False
     )
@@ -54,16 +54,16 @@ class YoutubeMediaView(MethodView):
     def get(self, video_id: str) -> ResponseReturnValue:
         if not leniently_validate_youtube_id(video_id):
             return Response("Invalid video ID", status=400)
-        config: Configuration = current_app.config["PODCAST_CONFIG"]
+        config: ServiceConfig = current_app.config["PODCAST_SERVICE_CONFIG"]
         validated_video_id = validate_youtube_video_id(video_id, config)
         if validated_video_id is None:
             return Response("Video ID does not exist", status=400)
         audio_output_path = config.data_path / "audio" / f"{validated_video_id}.m4a"
-        if audio_output_path.exists():
+        if audio_output_path.exists() and audio_output_path.is_file():
             return send_file(audio_output_path)
         # We need a per-video_id lock here to prevent two requests from causing the same video to download twice
         with self.video_download_locks[validated_video_id]:
-            if audio_output_path.exists():
+            if audio_output_path.exists() and audio_output_path.is_file():
                 return send_file(audio_output_path)
             logging.info(f"Downloading audio from YouTube video {validated_video_id}")
             try:

--- a/src/podcastsponsorblock/views/youtuberssview.py
+++ b/src/podcastsponsorblock/views/youtuberssview.py
@@ -18,11 +18,7 @@ from flask import (
 from flask.views import MethodView
 
 from ..helpers import YoutubePlaylistEpisodeFeed, leniently_validate_youtube_id, escape_for_xml, get_itunes_artwork
-from ..models import EpisodeDetails, ServiceConfig, PodcastConfig, FeedOptions
-
-
-class FeedAuthor(TypedDict):
-    name: str
+from ..models import EpisodeDetails, ServiceConfig, FeedOptions
 
 
 class Image(TypedDict):
@@ -66,7 +62,6 @@ def generate_episode_entry(
     feed_entry.id(episode.id)
     feed_entry.title(episode.title)
     feed_entry.description(episode.description)
-    feed_entry.author(FeedAuthor(name=episode.author.name))
     feed_entry.published(episode.published_at)
     if generator_options.service_config.append_auth_param_to_resource_links:
         feed_entry.enclosure(
@@ -103,9 +98,6 @@ def populate_feed_generator(
     playlist_details = playlist_episode_feed.playlist_details
     feed_generator = FeedGenerator()
     feed_generator.title(escape_for_xml(playlist_details.title))
-    feed_generator.author(FeedAuthor(name=escape_for_xml(playlist_details.author.name)))
-    feed_generator.load_extension("podcast")
-    podcast_config = generator_options.podcast_config
     youtube_playlist_url = (
         f"https://www.youtube.com/playlist?{urlencode({'list': playlist_details.id})}"
     )
@@ -119,8 +111,11 @@ def populate_feed_generator(
             link=youtube_playlist_url,
         )
     )
+    feed_generator.load_extension("podcast")
+    podcast_config = generator_options.podcast_config
     # noinspection PyUnresolvedReferences
     podcast_feed_generator = feed_generator.podcast
+    podcast_feed_generator.itunes_author(playlist_details.author.name)
     if podcast_config is not None:
         if podcast_config.itunes_id is not None:
             try:

--- a/src/podcastsponsorblock/views/youtuberssview.py
+++ b/src/podcastsponsorblock/views/youtuberssview.py
@@ -83,7 +83,7 @@ def generate_episode_entry(
             )
         )
     else:
-        feed_entry.link(
+        feed_entry.enclosure(
             **Enclosure(
                 url=add_host(
                     # Apple podcasts requires the file extension

--- a/src/podcastsponsorblock/views/youtuberssview.py
+++ b/src/podcastsponsorblock/views/youtuberssview.py
@@ -33,6 +33,7 @@ class Image(TypedDict):
 class Link(TypedDict):
     href: str
     rel: str
+    type: Optional[str]
 
 
 class Enclosure(TypedDict):
@@ -125,8 +126,8 @@ def populate_feed_generator(
         )
     )
     podcast_config = generator_options.podcast_config
-    # noinspection PyUnresolvedReferences
     if podcast_config is not None:
+        # noinspection PyUnresolvedReferences
         podcast_feed_generator = feed_generator.podcast
         if podcast_config.language is not None:
             feed_generator.language(escape_for_xml(podcast_config.language))
@@ -156,6 +157,7 @@ def populate_feed_generator(
                     generator_options
                 ),
                 rel="self",
+                type="application/rss+xml"
             )
         )
     else:
@@ -166,6 +168,7 @@ def populate_feed_generator(
                     generator_options
                 ),
                 rel="self",
+                type="application/rss+xml"
             )
         )
     return feed_generator

--- a/src/podcastsponsorblock/views/youtuberssview.py
+++ b/src/podcastsponsorblock/views/youtuberssview.py
@@ -74,8 +74,6 @@ def generate_episode_entry(
     feed_entry.title(episode.title)
     feed_entry.description(episode.description)
     feed_entry.author(FeedAuthor(name=episode.author.name))
-    feed_entry.load_extension("media")
-    # noinspection PyUnresolvedReferences
     feed_entry.published(episode.published_at)
     if generator_options.service_config.append_auth_param_to_resource_links:
         feed_entry.enclosure(
@@ -86,7 +84,7 @@ def generate_episode_entry(
                     ),
                     generator_options
                 ),
-                type="audio/mp4",
+                type="audio/x-m4a",  # Apple podcasts requires this instead of audio/mp4
                 length="0"
             )
         )
@@ -96,7 +94,7 @@ def generate_episode_entry(
                 url=add_host(
                     url_for("youtube_media_view", video_id=episode.id), generator_options
                 ),
-                type="audio/mp4",
+                type="audio/mp4",  # Apple podcasts requires this instead of audio/mp4
                 length="0"
             )
         )

--- a/src/podcastsponsorblock/views/youtuberssview.py
+++ b/src/podcastsponsorblock/views/youtuberssview.py
@@ -17,8 +17,8 @@ from flask import (
 )
 from flask.views import MethodView
 
-from ..helpers import YoutubePlaylistEpisodeFeed, leniently_validate_youtube_id, escape_for_xml
-from ..models import EpisodeDetails, ServiceConfig, PodcastConfig
+from ..helpers import YoutubePlaylistEpisodeFeed, leniently_validate_youtube_id, escape_for_xml, get_itunes_artwork
+from ..models import EpisodeDetails, ServiceConfig, PodcastConfig, FeedOptions
 
 
 class FeedAuthor(TypedDict):
@@ -46,18 +46,11 @@ class Thumbnail(TypedDict):
     url: str
 
 
-@dataclass
-class GeneratorOptions:
-    service_config: ServiceConfig
-    podcast_config: Optional[PodcastConfig]
-    host: str
-
-
 def is_absolute(url: str) -> bool:
     return urlparse(url).netloc != ""
 
 
-def add_host(url: str, generator_options: GeneratorOptions) -> str:
+def add_host(url: str, generator_options: FeedOptions) -> str:
     host = generator_options.host
     for allowed_host in generator_options.service_config.trusted_hosts:
         # noinspection HttpUrlsUsage
@@ -67,7 +60,7 @@ def add_host(url: str, generator_options: GeneratorOptions) -> str:
 
 
 def generate_episode_entry(
-    episode: EpisodeDetails, generator_options: GeneratorOptions
+    episode: EpisodeDetails, generator_options: FeedOptions
 ) -> FeedEntry:
     feed_entry = FeedEntry()
     feed_entry.id(episode.id)
@@ -91,7 +84,7 @@ def generate_episode_entry(
         )
     else:
         feed_entry.link(
-            Enclosure(
+            **Enclosure(
                 url=add_host(
                     # Apple podcasts requires the file extension
                     url_for("youtube_media_view", video_id=f"{episode.id}.m4a"), generator_options
@@ -105,13 +98,14 @@ def generate_episode_entry(
 
 def populate_feed_generator(
     playlist_episode_feed: YoutubePlaylistEpisodeFeed,
-    generator_options: GeneratorOptions
+    generator_options: FeedOptions
 ) -> FeedGenerator:
     playlist_details = playlist_episode_feed.playlist_details
     feed_generator = FeedGenerator()
     feed_generator.title(escape_for_xml(playlist_details.title))
     feed_generator.author(FeedAuthor(name=escape_for_xml(playlist_details.author.name)))
     feed_generator.load_extension("podcast")
+    podcast_config = generator_options.podcast_config
     youtube_playlist_url = (
         f"https://www.youtube.com/playlist?{urlencode({'list': playlist_details.id})}"
     )
@@ -125,19 +119,22 @@ def populate_feed_generator(
             link=youtube_playlist_url,
         )
     )
-    podcast_config = generator_options.podcast_config
+    # noinspection PyUnresolvedReferences
+    podcast_feed_generator = feed_generator.podcast
     if podcast_config is not None:
-        # noinspection PyUnresolvedReferences
-        podcast_feed_generator = feed_generator.podcast
+        if podcast_config.itunes_id is not None:
+            try:
+                itunes_artwork_url = get_itunes_artwork(podcast_config.itunes_id)
+            except ValueError as exception:
+                logging.error("Failed to grab iTunes artwork", exception)
+            else:
+                podcast_feed_generator.itunes_image(itunes_artwork_url)
         if podcast_config.language is not None:
             feed_generator.language(escape_for_xml(podcast_config.language))
         if podcast_config.explicit is not None:
             podcast_feed_generator.itunes_explicit("yes" if podcast_config.explicit else "no")
         if podcast_config.itunes_category is not None:
             podcast_feed_generator.itunes_category(escape_for_xml(podcast_config.itunes_category))
-        if podcast_feed_generator.itunes_image() is None:
-            podcast_feed_generator.itunes_image("https://is1-ssl.mzstatic.com/image/thumb/Podcasts116/v4/aa/2a/c8/aa2ac8b3-8f4e-d921-7046-9a910be576b5/mza_11213880277667169347.jpg/3000x3000bb.jpg")
-    logging.info(feed_generator._FeedGenerator__rss_link)
     if podcast_config is not None and podcast_config.description is not None:
         feed_generator.subtitle(escape_for_xml(podcast_config.description))
     elif playlist_details.description is not None:
@@ -153,7 +150,7 @@ def populate_feed_generator(
     key=lambda episode_feed, generator_options: hashkey(episode_feed.playlist_details.id, generator_options.host),
 )
 def generate_rss_feed(
-    episode_feed: YoutubePlaylistEpisodeFeed, generator_options: GeneratorOptions
+    episode_feed: YoutubePlaylistEpisodeFeed, generator_options: FeedOptions
 ) -> str:
     logging.info(
         f"Generating RSS feed for YouTube playlist {episode_feed.playlist_details.id}"
@@ -170,19 +167,18 @@ class YoutubeRSSView(MethodView):
         playlist_id = service_config.aliases.get(playlist_id.lower(), playlist_id)
         if not leniently_validate_youtube_id(playlist_id):
             return Response("Invalid playlist ID", status=400)
+        feed_options = FeedOptions(service_config, None, request.host)
         try:
             episode_feed = YoutubePlaylistEpisodeFeed(
-                playlist_id=playlist_id, config=service_config
+                playlist_id=playlist_id, feed_options=feed_options
             )
         except ValueError:
             return Response("Playlist does not exist", status=400)
         podcast_config = service_config.podcast_configs.get(episode_feed.playlist_details.id)
-        if len(service_config.trusted_hosts) > 0:
-            return Response(
-                generate_rss_feed(episode_feed, GeneratorOptions(service_config, podcast_config, request.host)),
-                mimetype="application/rss+xml",
-            )
-        else:
-            return Response(
-                generate_rss_feed(episode_feed, GeneratorOptions(service_config, podcast_config, "")), mimetype="application/rss+xml"
-            )
+        feed_options.podcast_config = podcast_config
+        if len(service_config.trusted_hosts) < 1:
+            feed_options.host = ""
+        return Response(
+            generate_rss_feed(episode_feed, feed_options),
+            mimetype="application/rss+xml",
+        )

--- a/src/podcastsponsorblock/views/youtuberssview.py
+++ b/src/podcastsponsorblock/views/youtuberssview.py
@@ -1,7 +1,8 @@
 import logging
+from dataclasses import dataclass
 from urllib.parse import urlparse, urlencode
 from datetime import timedelta
-from typing import TypedDict
+from typing import TypedDict, Optional
 
 from cachetools import cached, TTLCache
 from cachetools.keys import hashkey
@@ -16,8 +17,8 @@ from flask import (
 )
 from flask.views import MethodView
 
-from ..helpers import YoutubePlaylistEpisodeFeed, leniently_validate_youtube_id
-from ..models import EpisodeDetails, Configuration
+from ..helpers import YoutubePlaylistEpisodeFeed, leniently_validate_youtube_id, escape_for_xml
+from ..models import EpisodeDetails, ServiceConfig, PodcastConfig
 
 
 class FeedAuthor(TypedDict):
@@ -34,20 +35,30 @@ class Link(TypedDict):
     rel: str
 
 
-class Enclosure(Link):
+class Enclosure(TypedDict):
+    url: str
     type: str
+    length: str
 
 
 class Thumbnail(TypedDict):
     url: str
 
 
+@dataclass
+class GeneratorOptions:
+    service_config: ServiceConfig
+    podcast_config: Optional[PodcastConfig]
+    host: str
+
+
 def is_absolute(url: str) -> bool:
     return urlparse(url).netloc != ""
 
 
-def add_host(url: str, host: str, config: Configuration) -> str:
-    for allowed_host in config.trusted_hosts:
+def add_host(url: str, generator_options: GeneratorOptions) -> str:
+    host = generator_options.host
+    for allowed_host in generator_options.service_config.trusted_hosts:
         # noinspection HttpUrlsUsage
         if allowed_host == f"https://{host}" or allowed_host == f"http://{host}":
             return allowed_host + url
@@ -55,7 +66,7 @@ def add_host(url: str, host: str, config: Configuration) -> str:
 
 
 def generate_episode_entry(
-    episode: EpisodeDetails, config: Configuration, host: str
+    episode: EpisodeDetails, generator_options: GeneratorOptions
 ) -> FeedEntry:
     feed_entry = FeedEntry()
     feed_entry.id(episode.id)
@@ -64,30 +75,29 @@ def generate_episode_entry(
     feed_entry.author(FeedAuthor(name=episode.author.name))
     feed_entry.load_extension("media")
     # noinspection PyUnresolvedReferences
-    feed_entry.media.thumbnail(Thumbnail(url=add_host(episode.icon_url, host, config)))
+    feed_entry.media.thumbnail(Thumbnail(url=add_host(episode.icon_url, generator_options)))
     feed_entry.published(episode.published_at)
-    if config.append_auth_param_to_resource_links:
-        feed_entry.link(
-            Enclosure(
-                href=add_host(
+    if generator_options.service_config.append_auth_param_to_resource_links:
+        feed_entry.enclosure(
+            **Enclosure(
+                url=add_host(
                     url_for(
-                        "youtube_media_view", video_id=episode.id, key=config.auth_key
+                        "youtube_media_view", video_id=episode.id, key=generator_options.service_config.auth_key
                     ),
-                    host,
-                    config,
+                    generator_options
                 ),
-                rel="enclosure",
                 type="audio/mp4",
+                length="0"
             )
         )
     else:
         feed_entry.link(
             Enclosure(
-                href=add_host(
-                    url_for("youtube_media_view", video_id=episode.id), host, config
+                url=add_host(
+                    url_for("youtube_media_view", video_id=episode.id), generator_options
                 ),
-                rel="enclosure",
                 type="audio/mp4",
+                length="0"
             )
         )
     return feed_entry
@@ -95,39 +105,55 @@ def generate_episode_entry(
 
 def populate_feed_generator(
     playlist_episode_feed: YoutubePlaylistEpisodeFeed,
-    config: Configuration,
-    host: str,
+    generator_options: GeneratorOptions
 ) -> FeedGenerator:
     playlist_details = playlist_episode_feed.playlist_details
     feed_generator = FeedGenerator()
-    feed_generator.title(playlist_details.title)
-    feed_generator.author(FeedAuthor(name=playlist_details.author.name))
+    feed_generator.title(escape_for_xml(playlist_details.title))
+    feed_generator.author(FeedAuthor(name=escape_for_xml(playlist_details.author.name)))
+    feed_generator.load_extension("podcast")
     youtube_playlist_url = (
         f"https://www.youtube.com/playlist?{urlencode({'list': playlist_details.id})}"
     )
-    feed_generator.link(Link(href=youtube_playlist_url, rel="alternate"))
     podcast_logo_url = playlist_episode_feed.logo
     if not is_absolute(podcast_logo_url):
-        podcast_logo_url = add_host(podcast_logo_url, host, config)
+        podcast_logo_url = add_host(podcast_logo_url, generator_options)
     feed_generator.image(
         **Image(
             url=podcast_logo_url,
             link=youtube_playlist_url,
         )
     )
-    feed_generator.subtitle(playlist_details.description or "No description available")
+    podcast_config = generator_options.podcast_config
+    # noinspection PyUnresolvedReferences
+    if podcast_config is not None:
+        podcast_feed_generator = feed_generator.podcast
+        if podcast_config.language is not None:
+            feed_generator.language(escape_for_xml(podcast_config.language))
+        if podcast_config.explicit is not None:
+            podcast_feed_generator.itunes_explicit("yes" if podcast_config.explicit else "no")
+        if podcast_config.itunes_category is not None:
+            podcast_feed_generator.itunes_category(escape_for_xml(podcast_config.itunes_category))
+        if podcast_feed_generator.itunes_image() is None:
+            podcast_feed_generator.itunes_image("https://is1-ssl.mzstatic.com/image/thumb/Podcasts116/v4/aa/2a/c8/aa2ac8b3-8f4e-d921-7046-9a910be576b5/mza_11213880277667169347.jpg/3000x3000bb.jpg")
+    feed_generator.link(Link(href=youtube_playlist_url, rel="alternate"))
+    if podcast_config is not None and podcast_config.description is not None:
+        feed_generator.subtitle(escape_for_xml(podcast_config.description))
+    elif playlist_details.description is not None:
+        feed_generator.subtitle(escape_for_xml(playlist_details.description))
+    else:
+        feed_generator.subtitle("No description available")
     feed_generator.id(playlist_details.id)
-    if config.append_auth_param_to_resource_links:
+    if generator_options.service_config.append_auth_param_to_resource_links:
         feed_generator.link(
             Link(
                 href=add_host(
                     url_for(
                         "youtube_rss_view",
                         playlist_id=playlist_details.id,
-                        key=config.auth_key,
+                        key=generator_options.service_config.auth_key,
                     ),
-                    host,
-                    config,
+                    generator_options
                 ),
                 rel="self",
             )
@@ -137,8 +163,7 @@ def populate_feed_generator(
             Link(
                 href=add_host(
                     url_for("youtube_rss_view", playlist_id=playlist_details.id),
-                    host,
-                    config,
+                    generator_options
                 ),
                 rel="self",
             )
@@ -148,38 +173,39 @@ def populate_feed_generator(
 
 @cached(
     TTLCache(maxsize=1024, ttl=timedelta(minutes=60).total_seconds()),
-    key=lambda episode_feed, _, host: hashkey(episode_feed.playlist_details.id, host),
+    key=lambda episode_feed, generator_options: hashkey(episode_feed.playlist_details.id, generator_options.host),
 )
 def generate_rss_feed(
-    episode_feed: YoutubePlaylistEpisodeFeed, config: Configuration, host: str
+    episode_feed: YoutubePlaylistEpisodeFeed, generator_options: GeneratorOptions
 ) -> str:
     logging.info(
         f"Generating RSS feed for YouTube playlist {episode_feed.playlist_details.id}"
     )
-    feed_generator = populate_feed_generator(episode_feed, config, host)
+    feed_generator = populate_feed_generator(episode_feed, generator_options)
     for episode in episode_feed:
-        feed_generator.add_entry(generate_episode_entry(episode, config, host))
+        feed_generator.add_entry(generate_episode_entry(episode, generator_options))
     return feed_generator.rss_str()
 
 
 class YoutubeRSSView(MethodView):
     def get(self, playlist_id: str) -> ResponseReturnValue:
-        config: Configuration = current_app.config["PODCAST_CONFIG"]
-        playlist_id = config.aliases.get(playlist_id.lower(), playlist_id)
+        service_config: ServiceConfig = current_app.config["PODCAST_SERVICE_CONFIG"]
+        playlist_id = service_config.aliases.get(playlist_id.lower(), playlist_id)
         if not leniently_validate_youtube_id(playlist_id):
             return Response("Invalid playlist ID", status=400)
         try:
             episode_feed = YoutubePlaylistEpisodeFeed(
-                playlist_id=playlist_id, config=config
+                playlist_id=playlist_id, config=service_config
             )
         except ValueError:
             return Response("Playlist does not exist", status=400)
-        if len(config.trusted_hosts) > 0:
+        podcast_config = service_config.podcast_configs.get(episode_feed.playlist_details.id)
+        if len(service_config.trusted_hosts) > 0:
             return Response(
-                generate_rss_feed(episode_feed, config, request.host),
-                mimetype="text/xml",
+                generate_rss_feed(episode_feed, GeneratorOptions(service_config, podcast_config, request.host)),
+                mimetype="application/rss+xml",
             )
         else:
             return Response(
-                generate_rss_feed(episode_feed, config, ""), mimetype="text/xml"
+                generate_rss_feed(episode_feed, GeneratorOptions(service_config, podcast_config, "")), mimetype="application/rss+xml"
             )

--- a/src/podcastsponsorblock/views/youtuberssview.py
+++ b/src/podcastsponsorblock/views/youtuberssview.py
@@ -80,7 +80,8 @@ def generate_episode_entry(
             **Enclosure(
                 url=add_host(
                     url_for(
-                        "youtube_media_view", video_id=episode.id, key=generator_options.service_config.auth_key
+                        # Apple podcasts requires the file extension
+                        "youtube_media_view", video_id=f"{episode.id}.m4a", key=generator_options.service_config.auth_key
                     ),
                     generator_options
                 ),
@@ -92,7 +93,8 @@ def generate_episode_entry(
         feed_entry.link(
             Enclosure(
                 url=add_host(
-                    url_for("youtube_media_view", video_id=episode.id), generator_options
+                    # Apple podcasts requires the file extension
+                    url_for("youtube_media_view", video_id=f"{episode.id}.m4a"), generator_options
                 ),
                 type="audio/mp4",  # Apple podcasts requires this instead of audio/mp4
                 length="0"

--- a/src/podcastsponsorblock/views/youtuberssview.py
+++ b/src/podcastsponsorblock/views/youtuberssview.py
@@ -76,7 +76,6 @@ def generate_episode_entry(
     feed_entry.author(FeedAuthor(name=episode.author.name))
     feed_entry.load_extension("media")
     # noinspection PyUnresolvedReferences
-    feed_entry.media.thumbnail(Thumbnail(url=add_host(episode.icon_url, generator_options)))
     feed_entry.published(episode.published_at)
     if generator_options.service_config.append_auth_param_to_resource_links:
         feed_entry.enclosure(

--- a/src/podcastsponsorblock/views/youtuberssview.py
+++ b/src/podcastsponsorblock/views/youtuberssview.py
@@ -32,7 +32,7 @@ class Image(TypedDict):
 
 class Link(TypedDict):
     href: str
-    rel: str
+    rel: Optional[str]
     type: Optional[str]
 
 
@@ -119,6 +119,7 @@ def populate_feed_generator(
     podcast_logo_url = playlist_episode_feed.logo
     if not is_absolute(podcast_logo_url):
         podcast_logo_url = add_host(podcast_logo_url, generator_options)
+    feed_generator.link(Link(href=youtube_playlist_url, rel=None, type=None))
     feed_generator.image(
         **Image(
             url=podcast_logo_url,
@@ -137,7 +138,7 @@ def populate_feed_generator(
             podcast_feed_generator.itunes_category(escape_for_xml(podcast_config.itunes_category))
         if podcast_feed_generator.itunes_image() is None:
             podcast_feed_generator.itunes_image("https://is1-ssl.mzstatic.com/image/thumb/Podcasts116/v4/aa/2a/c8/aa2ac8b3-8f4e-d921-7046-9a910be576b5/mza_11213880277667169347.jpg/3000x3000bb.jpg")
-    feed_generator.link(Link(href=youtube_playlist_url, rel="alternate"))
+    logging.info(feed_generator._FeedGenerator__rss_link)
     if podcast_config is not None and podcast_config.description is not None:
         feed_generator.subtitle(escape_for_xml(podcast_config.description))
     elif playlist_details.description is not None:
@@ -145,32 +146,6 @@ def populate_feed_generator(
     else:
         feed_generator.subtitle("No description available")
     feed_generator.id(playlist_details.id)
-    if generator_options.service_config.append_auth_param_to_resource_links:
-        feed_generator.link(
-            Link(
-                href=add_host(
-                    url_for(
-                        "youtube_rss_view",
-                        playlist_id=playlist_details.id,
-                        key=generator_options.service_config.auth_key,
-                    ),
-                    generator_options
-                ),
-                rel="self",
-                type="application/rss+xml"
-            )
-        )
-    else:
-        feed_generator.link(
-            Link(
-                href=add_host(
-                    url_for("youtube_rss_view", playlist_id=playlist_details.id),
-                    generator_options
-                ),
-                rel="self",
-                type="application/rss+xml"
-            )
-        )
     return feed_generator
 
 

